vllm/v1/worker/neuron_worker.py

```diff
@@ -0,0 +1,189 @@
+from vllm.v1.kv_cache_interface import KVCacheConfig, KVCacheSpec
+from vllm.v1.outputs import ModelRunnerOutput
+from vllm.v1.worker.neuron_model_runner import NeuronModelRunner
+
+logger = init_logger(__name__)
+
+
+def get_current_memory_usage(rank):
+
+    process = subprocess.Popen("neuron-monitor",
+                               shell=False,
+                               stdout=subprocess.PIPE,
+                               preexec_fn=os.setsid)
+
+    try:
+        outs, errs = process.communicate(timeout=3)
+    except subprocess.TimeoutExpired:
+        process.kill()
+        outs, errs = process.communicate()
+        runtime_data = json.loads(outs)['neuron_runtime_data']
+        hardware_info = json.loads(outs)['neuron_hardware_info']
+        if len(runtime_data) == 0:
+            memory_used = 0
+        else:
+            memory_used = runtime_data[0]['report']['memory_used'][
+                'neuron_runtime_used_bytes']['neuron_device']
+
+        total_memory = hardware_info[
+            'neuron_device_memory_size'] // hardware_info[
+                'neuroncore_per_device_count']
+    return memory_used, total_memory
+
+
+class NeuronWorker:
+
+    def __init__(
+        self,
+        vllm_config: VllmConfig,
+        local_rank: int,
+        rank: int,
+        distributed_init_method: str,
+        is_driver_worker: bool = False,
+    ):
+        self.is_driver_worker = is_driver_worker
+        self.vllm_config = vllm_config
+        self.model_config = vllm_config.model_config
+        self.cache_config = vllm_config.cache_config
+        self.lora_config = vllm_config.lora_config
+        self.load_config = vllm_config.load_config
+        self.parallel_config = vllm_config.parallel_config
+        self.scheduler_config = vllm_config.scheduler_config
+        self.device_config = vllm_config.device_config
+        self.speculative_config = vllm_config.speculative_config
+        self.prompt_adapter_config = vllm_config.prompt_adapter_config
+        self.observability_config = vllm_config.observability_config
+
+        self.parallel_config.rank = rank
+        self.local_rank = local_rank
+        self.rank = rank
+        self.distributed_init_method = distributed_init_method
+
+        if self.cache_config.cache_dtype == "auto":
+            self.cache_dtype = self.model_config.dtype
+        else:
+            self.cache_dtype = STR_DTYPE_TO_TORCH_DTYPE[
+                self.cache_config.cache_dtype]
+
+        if self.model_config.trust_remote_code:
+            # note: lazy import to avoid importing torch before initializing
+            from vllm.utils import init_cached_hf_modules
+            init_cached_hf_modules()
+
+        self.profiler = None
+
+        if self.model_config.seed is None:
+            self.model_config.seed = 0
+
+    @torch.inference_mode()
+    def determine_available_memory(self) -> int:
+        self.model_runner.profile_run()
+        for _ in range(10):
```

Comment:
This loop was a hack to retry the invocation of `neuron-monitor`, which was failing frequently when there were multiple workers running. We should find a cleaner way to handle these retries (or at least add a note to follow up on this)

Reply:
added a note here.